File: src/main/java/oscar/cp/constraints/Le.java
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.cp.constraints;

import oscar.algo.Inconsistency;
import oscar.cp.core.CPPropagStrength;
import oscar.cp.core.Constraint;
import oscar.cp.core.variables.CPIntVar;
import oscar.cp.core.variables.CPVar;
import scala.collection.Iterable;
import scala.jdk.javaapi.CollectionConverters;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * Less Than Constraint ( x < y)
 * @author Pierre Schaus pschaus@gmail.com
 */
public class Le extends Constraint {

	private CPIntVar x, y;

    /**
     * x < y
     * @param x
     * @param y
     */
	public Le(CPIntVar x, CPIntVar y) {
		super(x.store()," < ");
		this.x = x;
		this.y = y;
	}
	
	public Le(CPIntVar x, int v) {
		this(x, CPIntVar.apply(v, v, x.store()));
	}

	@Override
	public Iterable<CPVar> associatedVars() {
		List<CPVar> l = new LinkedList<>(Arrays.asList(x, y));
		return CollectionConverters.asScala(l);
	}

	@Override
	public void setup(CPPropagStrength l) throws Inconsistency {
		if (y.isBound()) {
			x.updateMax(y.min()-1);
			return;
		}
		
		// y > x
		s().post(new Gr(y,x));
	}
	
}

File: src/main/scala/oscar/cp/constraints/Le.scala
package oscar.cp.constraints

import oscar.cp.core.CPPropagStrength
import oscar.cp.core.Constraint
import oscar.cp.core.variables.CPIntVar
import oscar.cp.core.variables.CPVar

class Le(val x: CPIntVar, val y: CPIntVar) extends Constraint(x.store, " < ") {

  def this(x: CPIntVar, v: Int) = this(x, CPIntVar(x.store, v, v))

  override def associatedVars(): Iterable[CPVar] = List(x, y)

  override def setup(l: CPPropagStrength): Unit = {
    if (y.isBound) {
      x.updateMax(y.min - 1)
      return
    }
    // y > x
    s.post(new Gr(y, x))
  }
}
